starlite/handlers/http_handlers.py

```diff
@@ -214,16 +213,13 @@ async def handler(data: Any, plugins: List["SerializationPluginProtocol"], **kwa
         if isawaitable(data):
             data = await data
 
-        if is_dto_annotation and not isinstance(data, DTO):
-            data = return_annotation(**data) if isinstance(data, dict) else return_annotation.from_model_instance(data)
+        if is_dto_annotation:
+            data = data.transfer_instance
```

Comment:
this removes the functionality where we can type a handler with a dto as return type, and then return something different from the handler.

It is a strange pattern to support IMO - so i pulled it for simplicity.

E.g. we were supporting a pattern that will always result in type errors..

```py
def handler() -> dto.Factory[MyModel]:
   ...
   return MyModel()  # type error
```

Reply:
@starlite-api/maintainers ping

Reply:
previously, when DTOs were pydantic models, and we had `# type:ignore`'s everywhere anyway, it may not have been as much as an issue.

Reply:
well, the point was that we were automatically generating the DTO there i think, no?

Reply:
Yes that is the point. My issue is that the code the pattern supports is a type error.

I will add the handling back if the consensus is against me. @provinzkraut any thoughts?

Reply:
I agree with @peterschutt that we shouldn’t support this as it is currently. If we want to do an implicit conversion, we’ll have to put the dto type elsewhere. Maybe a parameter on the handler or something like that?

Reply:
We need to automatically generate DTOs to have feature parity with FastAPI. There you can define a `response model` and use it to return a response and generate docs. This is our equivalent. 

It can be done in a different pattern, as long as we keep parity.

Reply:
> It can be done in a different pattern, as long as we keep parity.

That's what I was thinking of. Maybe a `response_dto` kwarg on the route handler? That wouldn't mess with the typing and be more explicit.

Reply:
> We need to automatically generate DTOs to have feature parity with FastAPI. There you can define a `response model` and use it to return a response and generate docs. This is our equivalent.
> 
> It can be done in a different pattern, as long as we keep parity.

For clarity, I still expect that we'll support this pattern:

```python
@post()
def handler(data: DomainType) -> DomainType:
   return data
```

Where, if `DomainType` is supported by a plugin, we will generate a data transfer model internally on startup, and that transfer type will represent _all_ attributes of the domain model - so it is probably only useful in narrow circumstances.

And @provinzkraut, you are suggesting something like this:

```python
@post(data_dto=DataDTO, response_dto=ResponseDTO)
def handler(data: DomainType) -> DomainType:
    return data
```
Where in this case, we use `DataDTO.from_bytes(...).to_model()` to instantiate the `DomainType` instance that is injected into the handler, and `ResponseDTO.from_model(data).to_bytes()` with the response from the handler to generate the actual response that is sent over the wire.

This would be functionally equivalent to:

```python
@post()
def handler(data: DataDTO) -> ResponseDTO:
    domain_instance = data.to_model()
    return ResponseDTO.from_model(domain_instance)
```

Which saves the explicit conversion to/from the dto type in user-land.

I think that is a nice pattern, and is effectively doing the same as:

```python
def handler() -> dto.Factory[MyModel]:
   ...
   return MyModel()  # type error
```
... but without the type errors.

Reply:
That's great then, no issues from my end. We should support also i think specifying a dto_type as well to allow users to exclude fields etc. Or we can allow passing drop factory options via route handler kwargs or extra

Reply:
> And @provinzkraut, you are suggesting something like this:

Yes, that's exactly what I was thinking of!

> We should support also i think specifying a dto_type as well to allow users to exclude fields etc. Or we can allow passing drop factory options via route handler kwargs or extra

Would this not already be covered by e.g. `response_dto`, since the user could easily create a DTO with field exclusion from a factory?

```python
@post(response_dto=dto_factory("DomainTypeWithoutID", DomainType, exlude=["id"]))
def handler(data: DomainType) -> DomainType:
    return data
```

Although I'd argue the proper pattern to encourage here would be to define the DTO outside the handler so it's a static class that's available in the namespace.